libs/sdk-common/src/tonic_wrap/mod.rs

```diff
@@ -0,0 +1,82 @@
+use std::{error::Error, fmt::Display, future::Future};
+
+use log::debug;
+
+pub struct Status(pub tonic::Status);
+
+impl Display for Status {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "status: {:?}, message: {:?}, details: {:?}, metadata: {:?}, source: {:?}",
+            self.0.code(),
+            self.0.message(),
+            self.0.details(),
+            self.0.metadata(),
+            self.0.source(),
+        )
+    }
+}
+
+pub struct TransportError(pub tonic::transport::Error);
+
+impl Display for TransportError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "description: {:?}, source: {:?}",
+            self.0.to_string(),
+            self.0.source(),
+        )
+    }
+}
+
+pub async fn with_connection_fallback<T, M, F>(
+    main: M,
+    fallback: impl FnOnce() -> F,
```

Comment:
I tried this, if you want you can try yourself too.
I tried by passing in a client to the function and having a `Fn` that operates on the client. I got stuck on lifetimes. It would be the cleaner way though.

Reply:
I think we can use a macro here so we can use it like this:

```
retry! {
 client.lsp_list(request.clone()).await
}
```

I will give it a try

Reply:
It seems that this macro should do it:

```
#[macro_export]
macro_rules! retry_grpc {
    ($f:expr) => {{
      use std::error::Error;
      use log::debug;
      const BROKEN_CONNECTION_STRINGS: [&str; 3] = [
          "http2 error: keep-alive timed out",
          "connection error: address not available",
          "connection error: timed out",
      ];

      let res = $f;
      let result = match res {
          Ok(t) => Ok(t),
          Err(status) => {
            let mut retruned = Err(status.clone());
            if let Some(source) = status.source() {
              if let Some(error) = source.downcast_ref::<tonic::transport::Error>() {
                if error.to_string() == "transport error" {
                  if let Some(source) = error.source() {
                    if BROKEN_CONNECTION_STRINGS.contains(&source.to_string().as_str()) {
                      debug!("retry_grpc: initial call failed due to broken connection. Retrying.");
                      retruned = $f
                    }
                  }
                }
              }
            }
            retruned
          },
      };
      result
   }};
}
```

Then it can be used as:

```
retry_grpc! {
 client.lsp_list(request.clone()).await
}
```